src/streamlink/plugins/twitch.py

```diff
@@ -125,27 +125,37 @@
 LOW_LATENCY_MAX_LIVE_EDGE = 2
 
 
-class TwitchM3U8Parser(M3U8Parser):
-    def parse_extinf(self, value):
-        duration, title = super(TwitchM3U8Parser, self).parse_extinf(value)
-        if title and str(title).startswith("Amazon"):
-            self.state["ad"] = True
+class TwitchM3U8(M3U8):
+    def __init__(self):
+        super(TwitchM3U8, self).__init__()
+        self.dateranges_ads = []
 
-        return duration, title
 
+class TwitchM3U8Parser(M3U8Parser):
     def parse_tag_ext_x_twitch_prefetch(self, value):
         segments = self.m3u8.segments
         if segments:
             segments.append(segments[-1]._replace(uri=self.uri(value), prefetch=True))
 
+    def parse_tag_ext_x_daterange(self, value):
+        super(TwitchM3U8Parser, self).parse_tag_ext_x_daterange(value)
+        daterange = self.m3u8.dateranges[-1]
+        is_ad = (
+            daterange.classname == "twitch-stitched-ad"
+            or str(daterange.id or "").startswith("stitched-ad-")
+            or any([attr_key.startswith("X-TV-TWITCH-AD-") for attr_key in daterange.x.keys()])
+        )
+        if is_ad:
+            self.m3u8.dateranges_ads.append(daterange)
+
     def get_segment(self, uri):
         byterange = self.state.pop("byterange", None)
         extinf = self.state.pop("extinf", (0, None))
         date = self.state.pop("date", None)
         map_ = self.state.get("map")
         key = self.state.get("key")
         discontinuity = self.state.pop("discontinuity", False)
-        ad = self.state.pop("ad", False)
+        ad = any([self.m3u8.is_date_in_daterange(date, daterange) for daterange in self.m3u8.dateranges_ads])
```

Comment:
I stumbled across this pull request while looking at the related issue. I figured I'd look at the code to see what was going and then figured I'd leave a small piece of feedback. If this wasn't the type of feedback you were hoping for then please ignore me :)

 Here and throughout this diff we are using `any([... for ... in ...])` and `all([... for ... in ...])`, these instances can be replaced by `any(... for ... in ...)` and `all(... for ... in ...)`. By removing the inner `[ ]` we are no longer creating intermediate lists, instead we are feeding a generator directly into `any` and `all`. This means that in the best case scenario we are iterating over fewer values since `any` and `all` will exit once they encounter the first `True` or `False` values respectively.

Thank you for all of your work on this, really love the software and appreciate the work you put into it!

Reply:
> If this wasn't the type of feedback you were hoping for then please ignore me :)

No, I appreciate it, thanks. It's just my inexperience with Python. I thought this was part of the iterator syntax. And yes, you're totally right, we don't need to feed an iterator into a list constructor to gather all of its results and then get the list's iterator to check for any/all matching results. That is wasteful.